File: Sources/Olaf/UI/Model/NetworkLogInfo.swift
import Foundation

/// `.network` kategorili bir `LogEntry`'nin metadata'sını yapısal network bilgisine ayrıştırır.
/// (Metadata anahtarları `OlafNetwork.NetworkLogComposer` ile hizalıdır.)
struct NetworkLogInfo {
    let method: String?
    let url: String?
    let statusCode: Int?
    let durationMs: Int?
    let requestBytes: Int?
    let responseBytes: Int?
    let error: String?
    /// İstek tamamlanmadan iptal edildi (hata değil; `.info` seviyesinde loglanır).
    let cancelled: Bool
    /// Yanıt mock'tan üretildi (ağa çıkılmadı).
    let mocked: Bool
    let requestBody: String?
    let responseBody: String?
    /// `image/*` yanıt gövdesi (sınır altındaysa yakalanır) — detayda önizleme için.
    let responseImageData: Data?
    let requestHeaders: [(key: String, value: String)]
    let responseHeaders: [(key: String, value: String)]

    // Zamanlama kırılımı (`t.` öneki — NetworkLogComposer ile hizalı; toplanamadıysa nil).
    let dnsMs: Int?
    let connectMs: Int?
    let tlsMs: Int?
    let ttfbMs: Int?
    let protocolName: String?
    let reusedConnection: Bool?

    /// Detayda "Zamanlama" bölümü gösterilecek mi?
    var hasTimings: Bool {
        dnsMs != nil || connectMs != nil || tlsMs != nil || ttfbMs != nil
            || protocolName != nil || reusedConnection != nil
    }

    init?(entry: LogEntry) {
        guard entry.category == .network else { return nil }
        let m = entry.metadata
        method = m["method"]
        url = m["url"]
        statusCode = m["status"].flatMap(Int.init)
        durationMs = m["durationMs"].flatMap(Int.init)
        requestBytes = m["reqBytes"].flatMap(Int.init)
        responseBytes = m["respBytes"].flatMap(Int.init)
        error = m["error"]
        cancelled = m["cancelled"] == "true"
        mocked = m["mocked"] == "true"
        requestBody = m["requestBody"]
        responseBody = m["responseBody"]
        responseImageData = m["responseImageBase64"].flatMap { Data(base64Encoded: $0) }
        requestHeaders = m
            .filter { $0.key.hasPrefix("reqH.") }
            .map { (String($0.key.dropFirst(5)), $0.value) }
            .sorted { $0.0 < $1.0 }
        responseHeaders = m
            .filter { $0.key.hasPrefix("respH.") }
            .map { (String($0.key.dropFirst(6)), $0.value) }
            .sorted { $0.0 < $1.0 }
        dnsMs = m["t.dnsMs"].flatMap(Int.init)
        connectMs = m["t.connectMs"].flatMap(Int.init)
        tlsMs = m["t.tlsMs"].flatMap(Int.init)
        ttfbMs = m["t.ttfbMs"].flatMap(Int.init)
        protocolName = m["t.protocol"]
        reusedConnection = m["t.reused"].map { $0 == "true" }
    }

    /// URL'in path (+ query) kısmı; satırda kısa gösterim için.
    var path: String {
        guard let url, let comps = URLComponents(string: url) else { return url ?? "-" }
        var p = comps.path.isEmpty ? "/" : comps.path
        if let q = comps.query, !q.isEmpty { p += "?\(q)" }
        return p
    }

    /// URL host'u.
    var host: String {
        guard let url, let comps = URLComponents(string: url) else { return "" }
        return comps.host ?? ""
    }

    /// Mock editörünün önerdiği eşleşme kalıbı: host + path (query'siz) — aynı endpoint'in
    /// tüm çağrılarını yakalayacak kadar genel, başka endpoint'lere taşmayacak kadar özgül.
    var suggestedMockPattern: String {
        guard let url, let comps = URLComponents(string: url) else { return url ?? "" }
        return (comps.host ?? "") + (comps.path.isEmpty ? "/" : comps.path)
    }

    var isFailure: Bool {
        if error != nil { return true }
        if let code = statusCode { return code >= 400 }
        return false
    }
}

File: Android/olaf/src/main/java/com/olaf/ui/model/NetworkLogInfo.kt
package com.olaf.ui.model

import com.olaf.LogCategory
import com.olaf.LogEntry
import java.util.Base64

/**
 * Parses the metadata of a `network`-category [LogEntry] into structured info. The keys are the
 * ones `NetworkLogComposer` writes — the same set the iOS package uses.
 */
internal data class NetworkLogInfo(
    val method: String?,
    val url: String?,
    val statusCode: Int?,
    val durationMs: Long?,
    val requestBytes: Long?,
    val responseBytes: Long?,
    val error: String?,
    /** Cancelled before completing — not a failure; logged at INFO. */
    val cancelled: Boolean,
    /** Produced by a mock, so no network call was made. */
    val mocked: Boolean,
    val requestBody: String?,
    val responseBody: String?,
    /** Image response body, when it was captured under the size limit. */
    val responseImageBytes: ByteArray?,
    val requestHeaders: List<Pair<String, String>>,
    val responseHeaders: List<Pair<String, String>>,
    val dnsMs: Long?,
    val connectMs: Long?,
    val tlsMs: Long?,
    val ttfbMs: Long?,
    val protocolName: String?,
    val reusedConnection: Boolean?
) {

    /** Should the detail screen show a "Timing" section? */
    val hasTimings: Boolean
        get() = dnsMs != null || connectMs != null || tlsMs != null ||
            ttfbMs != null || protocolName != null || reusedConnection != null

    /** Path (plus query) — the short form shown in a list row. */
    val path: String
        get() {
            val raw = url ?: return "-"
            val withoutScheme = raw.substringAfter("://", raw)
            val index = withoutScheme.indexOf('/')
            return if (index < 0) "/" else withoutScheme.substring(index)
        }

    val host: String
        get() {
            val raw = url ?: return ""
            return raw.substringAfter("://", raw).substringBefore('/').substringBefore('?')
        }

    /**
     * Match pattern the mock editor suggests: host + path without the query — broad enough to
     * catch every call to the endpoint, narrow enough not to spill onto other endpoints.
     */
    val suggestedMockPattern: String
        get() = host + path.substringBefore('?')

    val isFailure: Boolean
        get() = error != null || (statusCode != null && statusCode >= 400)

    override fun equals(other: Any?): Boolean = this === other ||
        (other is NetworkLogInfo && url == other.url && statusCode == other.statusCode && method == other.method)

    override fun hashCode(): Int = (url?.hashCode() ?: 0) * 31 + (statusCode ?: 0)

    companion object {
        /**
         * Returns `null` when the entry isn't a captured network record. Message-only entries the
         * host logs into the `network` category have no `url` key, and must fall back to the plain
         * message row instead of rendering as an empty network row.
         */
        fun from(entry: LogEntry): NetworkLogInfo? {
            if (entry.category != LogCategory.Network) return null
            val metadata = entry.metadata
            metadata["url"] ?: return null

            return NetworkLogInfo(
                method = metadata["method"],
                url = metadata["url"],
                statusCode = metadata["status"]?.toIntOrNull(),
                durationMs = metadata["durationMs"]?.toLongOrNull(),
                requestBytes = metadata["reqBytes"]?.toLongOrNull(),
                responseBytes = metadata["respBytes"]?.toLongOrNull(),
                error = metadata["error"],
                cancelled = metadata["cancelled"] == "true",
                mocked = metadata["mocked"] == "true",
                requestBody = metadata["requestBody"],
                responseBody = metadata["responseBody"],
                responseImageBytes = metadata["responseImageBase64"]?.let {
                    runCatching { Base64.getDecoder().decode(it) }.getOrNull()
                },
                requestHeaders = metadata.headers("reqH."),
                responseHeaders = metadata.headers("respH."),
                dnsMs = metadata["t.dnsMs"]?.toLongOrNull(),
                connectMs = metadata["t.connectMs"]?.toLongOrNull(),
                tlsMs = metadata["t.tlsMs"]?.toLongOrNull(),
                ttfbMs = metadata["t.ttfbMs"]?.toLongOrNull(),
                protocolName = metadata["t.protocol"],
                reusedConnection = metadata["t.reused"]?.let { it == "true" }
            )
        }

        private fun Map<String, String>.headers(prefix: String): List<Pair<String, String>> =
            filterKeys { it.startsWith(prefix) }
                .map { (key, value) -> key.removePrefix(prefix) to value }
                .sortedBy { it.first.lowercase() }
    }
}
